vllm/v1/core/scheduler.py

```diff
@@ -238,21 +234,20 @@ def schedule(self) -> "SchedulerOutput":
 
         # Next, schedule the WAITING requests.
         if not preempted_reqs:
-            # NOTE: We uses num_to_skip to determine
-            # which guided request within the waiting queue to skip
-            # over if the FSM of said request are yet to be ready.
-            num_to_skip: int = 0
-            while num_to_skip < len(self.waiting) and token_budget > 0:
+            # Use a temporary deque to collect requests that need to be skipped
+            # and put back at the head of the waiting queue later
+            still_waiting: deque[Request] = deque()
+            while self.waiting and token_budget > 0:
                 if len(self.running) == self.max_num_running_reqs:
                     break
 
-                request = self.waiting[num_to_skip]
+                request = self.waiting.popleft()
```

Comment:
There are some cases where a request gets lost. Check the code in between this pop and where it gets added to the running queue.